include/zephyr/bluetooth/audio/hap.h

```diff
@@ -0,0 +1,394 @@
+	 * @param start_index Preset start index that has been deleted.
+	 * @param end_index Preset end index that has been deleted.
+	 */
+	void (*remove)(struct bt_hap_harc *harc, uint8_t start_index, uint8_t end_index);
+
+	/**
+	 * @brief Callback function for preset availability notifications.
+	 *
+	 * The callback is called when the preset availability change is notified by the remote
+	 * server.
+	 *
+	 * @param harc Controller instance pointer.
+	 * @param index Preset index.
+	 * @param available True if available, false otherwise.
+	 */
+	void (*available)(struct bt_hap_harc *harc, uint8_t index, bool available);
+
+	/**
+	 * @brief Preset list update complete.
+	 *
+	 * Function called after all the preset list changes have been submitted.
+	 * User might apply the updated list to the application.
+	 *
+	 * @param harc Controller instance pointer.
+	 */
+	void (*commit)(struct bt_hap_harc *harc);
+
+	/**
+	 * @brief Preset get callback function.
+	 *
+	 * The callback is called to retrieve the stored preset record.
+	 *
+	 * @param harc Controller instance pointer.
+	 * @param index Preset index.
+	 * @param[out] record Preset record object to fill in.
+	 */
+	int (*get)(struct bt_hap_harc *harc, uint8_t index, struct bt_has_preset_record *record);
+};
+
+/**
+ * @brief Registers Hearing Aid Remote Controller Preset List callbacks.
+ *
+ * @param cb The callback structure.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_hap_harc_preset_cb_register(const struct bt_hap_harc_preset_cb *cb);
+
+/**
+ * @brief Get information for the profile connection.
+ *
+ * @param harc      Controller instance pointer.
+ * @param[out] info Info object.
+ *
+ * @return Zero on success or (negative) error code on failure.
+ */
+int bt_hap_harc_info_get(struct bt_hap_harc *harc, struct bt_hap_harc_info *info);
+
+/**
+ * @brief Get the connection pointer of a Hearing Aid Remote Controller instance
+ *
+ * Get the Bluetooth connection pointer of a Hearing Aid Remote Controller instance.
+ *
+ * @param harc Controller instance pointer.
+ * @param[out] conn Connection pointer.
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_hap_harc_conn_get(const struct bt_hap_harc *harc, struct bt_conn **conn);
+
+/**
+ * @brief Procedure complete result callback.
+ *
+ * @param err 0 in case of success or negative errno or ATT Error code in case of error.
+ * @param params Parameters passed in by the user.
+ */
+typedef void (*bt_hap_harc_complete_func_t)(int err, void *params);
+
+/**
+ * @brief Procedure status callback.
+ *
+ * @param harc Controller instance pointer.
+ * @param err 0 in case of success or negative errno or ATT Error code in case of error.
+ * @param params Procedure parameters.
+ */
+typedef void (*bt_hap_harc_status_func_t)(struct bt_hap_harc *harc, int err, void *params);
+
+/** @brief Read Preset parameters */
+struct bt_hap_harc_preset_read_params {
+	/** Complete result callback */
+	bt_hap_harc_complete_func_t complete;
```

Comment:
Providing the callback in the parameters is not necessarily better or worse than the approach we've otherwise taken with most profiles, but it's certainly different :)

It has some advantages, but also disadvantages. 

Can you elaborate on why you went with this direction, rather than the one we've more commonly used in LE Audio?

Reply:
Yes, the reason is quite important from application developer perspective.
Look at the following pseudo code:

`my_header.h`
```
/* Callback to the application */
struct my_callbacks {
   void (*value_cb)(struct my_inst* inst, int err, void *value);
};

/* Read the value */
int read_value(struct my_inst* inst);
```
`my_source.c`
```
/* Value notification callback */
void on_notification(struct my_inst* inst, void *value)
{
    cbs->value_cb(inst, 0, value);
}

void on_read_response(struct my_inst* inst, int err, void *value)
{
    cbs->value_cb(inst, 0, value);
}

int read_value(struct my_inst* inst)
{
   if (busy) {
         return -EBUSY;
   }
}
```
`my_app.c`
```
static k_sem value_read;
void value_cb()
{
  /* No one knows if it's the notification callback or read response callback */
  k_sem_give(value_read);
}

int main()
{
     int err;

     err = read_value(inst);
     /* err is 0 */
     k_sem_take(value_read);
       err = read_value(inst);
      /* err is -EBUSY because the callback was called on notification, not read response */  
}
```

I've seen such issue e.g. in `struct bt_micp_mic_ctlr_cb.mute`

Thus I added the separate callback parameter so that the caller knows when the operation is complete an it can start another one.